### ISSUE TITLE
Should "revoke" delete data from relying party service?

### ISSUE THREAD

**Issue by groovecoder** [mozilla/fxa#3744]
## Description

When someone revokes Monitor from their FxA, we DO NOT implicitly delete their data from Monitor. Should we?

### Steps to reproduce

As described in https://bugzilla.mozilla.org/show_bug.cgi?id=1598314 ...

1. Go to: https://monitor.firefox.com;
2. Login using the same account from the FxA;
3. Go to Manage account;
4. Tap on the "Show" button from the "Devices & apps" section;
5. Revoke the Firefox Monitor service;
6. Return to the Firefox Monitor page;
7. Refresh the page.
8. On our new branch, you will be automatically signed out

9. If you sign in again, we get a fresh token, and you get back into your existing Monitor account

Should we instead delete all of a user's Monitor data when they revoke the Monitor app?

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/FXA-844)
┆Issue Number: FXA-844

**Comment by 6a68** [comment]
My feeling is no, revoking the Monitor token via the FxA settings page should not be taken to mean, "I want to delete my Monitor account." But I'm not 100% sure, so marking this as needs-discussion. @rfk, any thoughts?

A related issue has come up before: #1203 asked whether revoking tokens on the FxA website should cause users to get immediately logged out of RPs like Monitor (answer: FxA doesn't support that).

I don't know if the OAuth/OIDC standards take manual token deletion to mean anything different than other token revocation scenarios, like password reset or simple token expiration. I have a hunch this is discussed in the specs somewhere, though I can't seem to find anything in token revocation spec [RFC 7009](https://tools.ietf.org/html/rfc7009).

If a user deletes their Firefox Account, it's clear that we want to notify RPs to delete data associated with that Firefox Account (#472), and it looks like this has already been set up for Monitor (https://github.com/mozilla/blurts-server/issues/710).

On the other hand, when a user changes their password, their access tokens should be revoked. We definitely wouldn't want a user's Monitor data to get deleted just because they've changed their password.

I think manually revoking tokens might fall somewhere in between, but I'm doubtful we'd want to treat it like an account deletion.

Related: maybe it would be helpful for FxA to write some general relying party explainer docs to cover this confusing overlap between FxA and RP accounts / tokens / sessions?

**Comment by rfk** [comment]
> revoking the Monitor token via the FxA settings page should not be taken to mean,
> "I want to delete my Monitor account."

I agree, but also...if that's not what it means, what exactly *does* the user mean?

In practice right now we are interpreting it as "prevent the Monitor service from getting updated account info, and lazily sign me out of monitor.firefox.com when I next try to access that website". But that seems too narrowly specific.

My personal take is that revoking the Monitor service through FxA should mean "disabling" the Monitor service on my account. That includes signing me out of the website, and also disabling future scans and emails and other background tasks that Monitor might be doing on my behalf.

That's not the same as deleting all my data in the Monitor service. For example, if I later decide I want to re-active my use of the service, I think I'd be surprised to sign in again and discover that Monitor had discarded all my settings, extra email addresses, etc.

> Related: maybe it would be helpful for FxA to write some general relying party explainer docs to
> cover this confusing overlap between FxA and RP accounts / tokens / sessions?

💯; I have many partially-formed thoughts on this that I would love to work with someone to turn into a doc. (Past experience has taught me that the "work with someone" part is important for my own accountability on getting these things done).

**Comment by groovecoder** [comment]
> On the other hand, when a user changes their password, their access tokens should be revoked. We definitely wouldn't want a user's Monitor data to get deleted just because they've changed their password.

I didn't know that, and that's a great point. Agree - we should automatically delete a user's Monitor data when they change their FxA password!

> My personal take is that revoking the Monitor service through FxA should mean "disabling" the Monitor service on my account. That includes signing me out of the website, and also disabling future scans and emails and other background tasks that Monitor might be doing on my behalf.

Another great point. That will take some work on our side to mark a subscriber as "disabled" and to think thru the UX implications of that. The only background task we do right now is to send breach alerts, so we could disable that.

**Comment by groovecoder** [comment]
Maybe this issue is just to send a webhook call when clients/tokens are revoked? And to give RPs general advice on what to do with those?

**Comment by 6a68** [comment]
> Maybe this issue is just to send a webhook call when clients/tokens are revoked? And to give RPs general advice on what to do with those?

TL;DR: After thinking it over, it's probably best to backlog or close this issue, in favor of a commitment to write some better RP docs (@rfk, I'm happy to assist with this). Monitor should just follow OAuth and treat a revoked token as "user is logged out", not as any kind of special deactivated state.

_Extended meditation on identity systems and opportunity cost follows. Feel free to skim or skip._

In OAuth terms, this bug is a wontfix. OAuth is a standard for an identity provider (here, FxA) to broker a third party's (here, Monitor) access to user data. When a user signs in to Monitor with FxA, they are telling FxA that Monitor has permission to access some of their personal data that's controlled/brokered by FxA (user data in the profile scope). In OAuth terms, token revocation just means, Monitor must re-ask the user for this permission before accessing the user's data again.

What we're discussing in this issue is a fundamental change to how FxA works, adding a new "really, really logout" flow, and a new "deactivated" zombie account state somewhere in between "logged out" and "account deleted", that isn't covered in the existing OAuth standards. There are many large UX and UR questions here, including whether users care or need this third state; how we could explain this third state to them; how many users ever find or click that "Revoke" button; and what higher-value projects would have to be dropped to figure out the details. Further, because this proposed new "deactivate" flow lives on the settings page, the feature would only make sense if we asked _every_ RP to add support for this new flow. There are many questions to consider: what does it mean to log back in from the deactivated state? What does it mean for a user to "deactivate" the website token, but not the mobile app token, or vice versa, if both website and mobile app access user data cached on the same RP server? Should RPs themselves provide a "deactivate" button? We'd essentially be rolling our own auth standard, which, as a former member of the Persona team, I can tell you is a difficult problem, and one that requires deep organizational support and long timelines. Hard to see this edge case as central enough to FxA's value proposition to justify the investment.

I think the best outcome here would be for me and @rfk to write up some docs in Q1 that clarify how OAuth token states map to user expectations (or, more specifically, how OAuth doesn't quite map to user expectations, like the funky edge case we're currently discussing). Concretely, FxA might want to rethink the "Revoke" and "Disconnect" button labels shown in the settings page. This might already have been covered in UX's new design work for the settings page; I'm not sure.

@groovecoder I hope this is an understandable, if not totally satisfying, conclusion :-P

**Comment by groovecoder** [comment]
That makes sense to me. I've got a PR open to fix our issue (https://github.com/mozilla/blurts-server/pull/1452) and I think what I'll suggest for Monitor is:

1. If we get a `401` response to the request with our access token, we'll try to use our refresh token to get a fresh access token and use that to get the profile data
2. If we get a `400` response to the request with our access token, with `errno: 108`, inform the user that Monitor has been revoked from their Firefox Account, and prompt them to either:
  * Reconnect Monitor to their FxA (Trigger the oauth flow again)
  * Delete Firefox Monitor from their account

**Comment by rfk** [comment]
I don't disagree with anything @6a68 said above, but I wanted to add a few extra cents.

> In OAuth terms, this bug is a wontfix. OAuth is a standard for an identity provider (here, FxA) to
> broker a third party's (here, Monitor) access to user data.

It's a bit weird because Monitor is not a third party as far as the user is concerned, is part of the "Firefox Family" and is advertised as part of a bundle of services tied closely to your account. This will inevitably have an impact on user expectations (even if we don't really understand what those expectations are just yet).

> What we're discussing in this issue is a fundamental change to how FxA works, adding a new
> "really, really logout" flow, and a new "deactivated" zombie account state somewhere in between
> "logged out" and "account deleted", that isn't covered in the existing OAuth standards.

I'm not sure this zombie state would need to exist in FxA, and it *already* exists for Monitor when the user revokes the refresh token. Leaving it to Monitor to do something sensible in this state makes sense to me, although we could help by pushing out a notification to inform it that the revocation event occurred.

> I think the best outcome here would be for me and @rfk to write up some docs in Q1 that clarify
> how OAuth token states map to user expectations (or, more specifically, how OAuth doesn't quite
> map to user expectations, like the funky edge case we're currently discussing).

👍 sounds great!
